src/Cooked/PrintToBuiltinString.hs

```diff
@@ -0,0 +1,357 @@
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# HLINT ignore "Use camelCase" #-}
+{-# HLINT ignore "Use foldr" #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
+
+-- | Print all the types that occur on the 'TxInfo' to 'BuiltinString'. This is
+-- useful for debugging of validators. You probably do not want to use this in
+-- production code, as many of the functions in this module are wildly
+-- inefficient due to limitations of the 'BuiltinString' type.
+module Cooked.PrintToBuiltinString (PrintBS (..), printBSs, app_prec) where
+
+import Plutus.V2.Ledger.Api
+import qualified PlutusTx.AssocMap as PlMap
+import PlutusTx.Builtins
+import PlutusTx.Prelude
+
+-- | analogue of Haskell's 'Show' class for use in Plutus scripts.
+class PrintBS a where
+  -- | analogue of 'show'
+  {-# INLINEABLE printBS #-}
+  printBS :: a -> BuiltinString
+  printBS x = printBSsPrec 0 x ""
+
+  -- | analogue of 'showsPrec'
+  {-# INLINEABLE printBSsPrec #-}
+  printBSsPrec :: Integer -> a -> BuiltinString -> BuiltinString
+  printBSsPrec _ x s = printBS x <> s
+
+-- | analogue of 'shows'
+{-# INLINEABLE printBSs #-}
+printBSs :: PrintBS a => a -> BuiltinString -> BuiltinString
+printBSs = printBSsPrec 0
+
+-- | Precedence of function applications
+{-# INLINEABLE app_prec #-}
+app_prec :: Integer
+app_prec = 10
+
+{-# INLINEABLE literal #-}
+literal :: BuiltinString -> BuiltinString -> BuiltinString
+literal = (<>)
+
+{-# INLINEABLE cat #-}
+cat :: [BuiltinString] -> BuiltinString -> BuiltinString
+cat [] = id -- we cannot use foldr here for some reason with the plutus compiler
+cat (x : xs) = literal x . cat xs
+
+-- | print with a surrounding parenthesis, if the boolean argument is true
+{-# INLINEABLE printBSParen #-}
+printBSParen :: Bool -> (BuiltinString -> BuiltinString) -> BuiltinString -> BuiltinString
+printBSParen False s = s
+printBSParen True s = literal "(" . s . literal ")"
+
+-- | print an application of a constructor to an argument
+{-# INLINEABLE application1 #-}
+application1 :: PrintBS a => Integer -> BuiltinString -> a -> BuiltinString -> BuiltinString
+application1 prec f x = printBSParen (app_prec < prec) $ literal f . literal " " . printBSsPrec (app_prec + 1) x
```

Comment:
```suggestion
application1 prec f x = printBSParen (app_prec <= prec) $ literal f . literal " " . printBSsPrec app_prec x
```
`<=` is necessary instead of `<` when deciding to print parenthesis in nested applications. 

Also, since `prec` reports the precedence of the calling context it isn't clear why adding (+1) is necessary when the context is actually an application.